ambry-replication/src/main/java/com/github/ambry/replication/ReplicaThread.java

```diff
@@ -348,6 +377,296 @@ protected Map<DataNodeId, List<RemoteReplicaInfo>> selectReplicas(Map<DataNodeId
+   * @param groupIdIterationCountMap {@link Map} map of group id to number of iterations
+   * @param allReplicasCaughtUpEarly {@link MutableBoolean} stores whether all replicas have caught up early
+   * @param maxIterationsReached {@link MutableBoolean} stores whether max iterations reached
+   */
+  void generateRemoteReplicaGroups(Map<Integer, List<RemoteReplicaInfo>> groupIdToRemoteReplicaMap,
+      Map<DataNodeId, Integer> standByNoProgressGroupIds, Map<Integer, RemoteReplicaGroup> inflightRemoteReplicaGroup,
+      Map<DataNodeId, Set<RemoteReplicaInfo>> standByNoProgressReplicaQueue,
+      Map<RemoteReplicaInfo, Long> remoteReplicaToThrottledTill, Map<Integer, Integer> groupIdIterationCountMap,
+      MutableBoolean allReplicasCaughtUpEarly, MutableBoolean maxIterationsReached) {
+
+    //check and process finished groups
+    processFinishedGroups(inflightRemoteReplicaGroup, remoteReplicaToThrottledTill);
+    logger.trace("Thread name: {} Finished processing finished groups", threadName);
+    //check if we should create new groups, return other wise
+    if (!shouldCreateNewRemoteReplicaGroups(inflightRemoteReplicaGroup, groupIdIterationCountMap,
+        allReplicasCaughtUpEarly, maxIterationsReached)) {
+      logger.trace("Thread name: {} Not creating any new groups", threadName);
+      return;
+    }
+
+    logger.trace("Thread name: {} Trying to create new active groups", threadName);
+    //now we create all active groups and add stand by no progress replicas to queue to be processed later
+    for (Map.Entry<Integer, List<RemoteReplicaInfo>> entry : groupIdToRemoteReplicaMap.entrySet()) {
+      Integer groupId = entry.getKey();
+      logger.trace("Thread name: {} Processing for group id {}", threadName, groupId);
+      List<RemoteReplicaInfo> remoteReplicas = entry.getValue();
+      DataNodeId remoteNode = remoteReplicas.get(0).getReplicaId().getDataNodeId();
+
+      //get stand by group id for the data node
+      Integer standbyReplicasNoProgressReplicaGroupId = standByNoProgressGroupIds.get(remoteNode);
+
+      List<RemoteReplicaInfo> activeReplicas = new ArrayList<>();
+      List<RemoteReplicaInfo> standbyReplicasWithNoProgress = new ArrayList<>();
+
+      //generate active and standby replicas with no progress
+      getActiveAndStandByNoProgressReplicas(activeReplicas, standbyReplicasWithNoProgress, remoteReplicas,
+          remoteReplicaToThrottledTill, standbyReplicasNoProgressReplicaGroupId, groupId, inflightRemoteReplicaGroup);
+      logger.trace("Thread name: {} Active replicas are {}, standby Replicas with no progress are {}", threadName,
+          activeReplicas, standbyReplicasWithNoProgress);
+
+      //if the current group is not in iteration, and there are active replicas, we will start a new group
+      //and increase iteration count if we are able to create
+      if (!inflightRemoteReplicaGroup.containsKey(groupId)) {
```

Comment:
instead of these maps and ids, please use an abstraction that can group together relevant info. For eg, this could be the state variable of an abstraction like IN_FLIGHT. Please cut back on the maps.

Reply:
refactored it.